07_memory/userspace/memory.c

```diff
@@ -0,0 +1,231 @@
+}
+
+unsigned long long get_av_meas_time(const int test_count)
+{
+	struct timespec t_begin = {0};
+	struct timespec t_end = {0};
+
+	unsigned long long time_arr[test_count];
+
+	for (int i = 0; i < test_count; i++)
+	{
+		clock_gettime(CLOCK_MONOTONIC, &t_begin);
+		clock_gettime(CLOCK_MONOTONIC, &t_end);
+		time_arr[i] = diff_tm_ns(&t_begin, &t_end);
+	}
+	return avarage(time_arr, test_count);
+}
+
+void do_mem_range_test(struct time_test_obj *test_obj, const int max_range, 
+			const int test_count, unsigned long long meas_time, 
+			int plus_delta) 
+{
+	unsigned long long without_delay_alloc = 0;
+	unsigned long long without_delay_deall = 0;
+
+	printf("************TEST %s***********\n", test_obj->fun_name);
+
+	printf("\n******2^X + %i*********\n", plus_delta);
+	for (int i = 0; i < max_range; i++) {
+				
+		unsigned long long buf_size = POW2(i) + (unsigned long long)plus_delta;
+		
+		aldeal_get_time_stats(test_obj, buf_size, test_count);
+
+		if (test_obj->avg_allocate == ULLONG_MAX) {
+			printf("Size: 2^%i + %i. NOT Allocated.\n", 
+							i, plus_delta);
+			break;	
+		} 
+
+		without_delay_alloc = test_obj->avg_allocate > meas_time?
+					test_obj->avg_allocate - meas_time:0;
+		without_delay_deall = test_obj->avg_deallocate > meas_time?
+					test_obj->avg_deallocate - meas_time:0;
+
+		printf("Size: 2^%i + %i. Allocation TIME: %llu (%llu) ns." 
+				" | Free Time: %llu (%llu) ns\n", i, plus_delta, 
+				test_obj->avg_allocate, without_delay_alloc, 
+				test_obj->avg_deallocate, without_delay_deall);
+	}
+	
+	printf("************TEST END***********\n\n");
+}
+
+
+
+void *all_malloc(unsigned long long count, unsigned long long size_of, unsigned long long *time)
+{
+	struct timespec t_begin = {0};
+	struct timespec t_end = {0};
+	char *buf = NULL;
+
+	clock_gettime(CLOCK_MONOTONIC, &t_begin);
+	buf = malloc(count * size_of);
+	clock_gettime(CLOCK_MONOTONIC, &t_end);
+
+	*time = diff_tm_ns(&t_begin, &t_end);
+
+	return buf;
+}
+
+void *all_calloc(unsigned long long count, unsigned long long size_of, unsigned long long *time)
+{
+	struct timespec t_begin = {0};
+	struct timespec t_end = {0};
+	char *buf = NULL;
+
+	clock_gettime(CLOCK_MONOTONIC, &t_begin);
+	buf =calloc(count, size_of);
+	clock_gettime(CLOCK_MONOTONIC, &t_end);
+
+	*time = diff_tm_ns(&t_begin, &t_end);
+
+	return buf;
+}
+
+void *all_alloca(unsigned long long count, unsigned long long size_of, unsigned long long *time)
+{	
+	struct timespec t_begin = {0};
+	struct timespec t_end = {0};
+	char *buf = NULL;
+	struct rlimit limit = {0};
+	int cur_stack_usage = 10000; //approx magic number
+	
+	getrlimit(RLIMIT_STACK, &limit);
+
+	if ((count * size_of) > (limit.rlim_cur - cur_stack_usage)) {
+		*time = 0;
+		return NULL;
+	}
+	
+	clock_gettime(CLOCK_MONOTONIC, &t_begin);
+	buf = alloca(count * size_of);
+	clock_gettime(CLOCK_MONOTONIC, &t_end);
+
+	*time = diff_tm_ns(&t_begin, &t_end);
+
+	return malloc(1); // we need return rubbish for pass NULL check
```

Comment:
What is the reason for this additional malloc(1). Can You use buf as result?

Reply:
Yes, as rubbish we can use buf result, but in case outside dereference we will got segfault or just access to random stack cell, that's why I decided to use this malloc.